File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/FlakeHelper.java
package com.github.worldoneo.ciphelet.connector;

public final class FlakeHelper {
    private static final char[] chars = "ABCDEFGH=JKLMN!PQRSTUVWXYZabcdefghijk/mnopqrstuvwxyz1234567890-+".toCharArray();

    public static String humanIDFromFlake(long flake) {
        StringBuilder stringBuilder = new StringBuilder();
        while (flake > 0) {
            stringBuilder.append(chars[(int) flake & 0x3F]);
            flake >>= 6;
        }
        return stringBuilder.toString();
    }

    public static long flakFromHumanID(String humanID) {
        long flake = 0L;
        for (byte b : new StringBuilder(humanID).reverse().toString().getBytes()) {
            flake <<= 6;
            for (int i = 0; i < chars.length; i++)
                if (chars[i] == b)
                    flake += i;
        }
        return flake;
    }

    public static long timeStampOfFlake(long flake) {
        return flake >> 21;
    }
}

File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/FlakeHelper.kt
package com.github.worldoneo.ciphelet.connector

object FlakeHelper {
    private val chars = "ABCDEFGH=JKLMN!PQRSTUVWXYZabcdefghijk/mnopqrstuvwxyz1234567890-+".toCharArray()
    fun humanIDFromFlake(flake: Long): String {
        var vf = flake
        val stringBuilder = StringBuilder()
        while (vf > 0) {
            stringBuilder.append(chars[vf.toInt() and 0x3F])
            vf = vf shr 6
        }
        return stringBuilder.toString()
    }

    fun flakFromHumanID(humanID: String?): Long {
        var flake = 0L
        for (b in StringBuilder(humanID!!).reverse().toString().toByteArray()) {
            flake = flake shl 6
            for (i in chars.indices) if (chars[i].toByte() == b) flake += i.toLong()
        }
        return flake
    }

    fun timeStampOfFlake(flake: Long): Long {
        return flake shr 21
    }
}